### PR TITLE
HIVE-23432 : Add Ranger Replication Metrics

### PR DESCRIPTION
**Pull request by aasha**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java [hunk 1]
```diff
@@ -68,4 +68,9 @@ public void endLog(String lastReplId) {
     (new BootstrapDumpEnd(dbName, tableSeqNo, functionSeqNo, dumpDir, lastReplId))
             .log(LogTag.END);
   }
+
+  @Override
+  public void endLog(long totalCount) {
+    //Do Nothing
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java:74]: Why not to provide default no-op version in base class itself and override only when required? Or even better, use generics for value top, that way only one version of endLog would be required?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/IncrementalDumpLogger.java [hunk 2]
```diff
@@ -60,4 +60,9 @@ public void eventLog(String eventId, String eventType) {
   public void endLog(String lastReplId) {
     (new IncrementalDumpEnd(dbName, eventSeqNo, dumpDir, lastReplId)).log(LogTag.END);
   }
+
+  @Override
+  public void endLog(long totalCount) {
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/IncrementalDumpLogger.java:65]: Same here

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/RangerDumpLogger.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  @Override
+  public void endLog(String lastReplId) {
+    //Do nothing for Ranger
+  }
+
+  @Override
+  public void endLog(long count) {
+    (new RangerDumpEnd(dbName, count, dumpDir))
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/RangerDumpLogger.java:51]: nit: Can be in single line

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  @JsonProperty
+  private String dbName;
+
+  @SuppressFBWarnings("URF_UNREAD_FIELD")
+  @JsonProperty
+  private Long dumpStartTime;
+
+  public RangerDumpBegin(String dbName) {
+    this.dbName = dbName;
+    this.dumpStartTime = System.currentTimeMillis() / 1000;
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java:40]: What's the idea behind this division by 1000?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.hadoop.hive.ql.parse.repl.ReplState;
+import org.codehaus.jackson.annotate.JsonProperty;
+
+/**
+ * RangerDumpBegin.
+ *
+ * ReplState to define Ranger Dump Start.
+ **/
+public class RangerDumpBegin extends ReplState {
+  @SuppressFBWarnings("URF_UNREAD_FIELD")
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java:30]: "UrF: Unread field (URF_UNREAD_FIELD)
This field is never read.  Consider removing it from the class."
The description says so, why is this required?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpEnd.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.hadoop.hive.ql.parse.repl.ReplState;
+import org.codehaus.jackson.annotate.JsonProperty;
+
+/**
+ * RangerDumpEnd.
+ *
+ * ReplState to define Ranger Dump End.
+ **/
+public class RangerDumpEnd extends ReplState {
+  @SuppressFBWarnings("URF_UNREAD_FIELD")
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpEnd.java:30]: URF_UNREAD_FIELD: Same here and in other fields as well.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/log/RangerLoadLogger.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  public RangerLoadLogger(String sourceDbName, String targetDbName, String dumpDir, long estimatedNumPolicies) {
+    this.sourceDbName = sourceDbName;
+    this.targetDbName = targetDbName;
+    this.estimatedNumPolicies = estimatedNumPolicies;
+    this.dumpDir = dumpDir;
+  }
+
+  @Override
+  public void startLog() {
+    (new RangerLoadBegin(sourceDbName, targetDbName, estimatedNumPolicies)).log(LogTag.RANGER_LOAD_START);
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/log/RangerLoadLogger.java:45]: nit: Remove extra brackets

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/log/RangerLoadLogger.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  @Override
+  public void endLog(String lastReplId) {
+    //Do nothing for Ranger
+  }
+
+  @Override
+  public void endLog(long count) {
+    (new RangerLoadEnd(sourceDbName, targetDbName, count, dumpDir))
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/log/RangerLoadLogger.java:55]: nit: Remove extra brackets and accommodate in single line

File: ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestRangerDumpTask.java [hunk 9]
```diff
@@ -112,4 +117,33 @@ public void testSuccessNonEmptyRangerPolicies() throws Exception {
...
+  @Test
+  public void testSuccessRangerDumpMetrics() throws Exception {
+    Logger logger = Mockito.mock(Logger.class);
+    Whitebox.setInternalState(ReplState.class, logger);
+    RangerExportPolicyList rangerPolicyList = new RangerExportPolicyList();
+    rangerPolicyList.setPolicies(new ArrayList<RangerPolicy>());
+    Mockito.when(mockClient.exportRangerPolicies(Mockito.anyString(), Mockito.anyString(), Mockito.anyString()))
+      .thenReturn(rangerPolicyList);
+    Mockito.when(conf.getVar(REPL_AUTHORIZATION_PROVIDER_SERVICE_ENDPOINT)).thenReturn("rangerEndpoint");
+    Mockito.when(conf.getVar(REPL_RANGER_SERVICE_NAME)).thenReturn("cm_hive");
```
Comment (pkumarsinha) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestRangerDumpTask.java:130]: change to hive

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java (hunk 1 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java:74]: Atleast one endLog has to be present. If you provide a implementation in the base class, there maybe a derived class which doesn't override.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/IncrementalDumpLogger.java (hunk 2 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/IncrementalDumpLogger.java:65]: Atleast one endLog has to be present. If you provide a implementation in the base class, there maybe a derived class which doesn't override.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java (hunk 4 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java:40]: Its printed in seconds. Thats the same with all metrics. The start time and endtime is shown in secs. Wanted to maintain consistency

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java (hunk 5 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java:30]: This is deserialised by the json deserializer. Its not unused. We don't need getter

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpEnd.java (hunk 6 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpEnd.java:30]: This is deserialised by the json deserializer. Its not unused. We don't need getter

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java (hunk 1 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java:74]: What I was referring to is:
Option 1: If you use generics you don't need extra endLog method and you can still define endLog as abstract, just the signature will change to 
Option 2: If you want separate endLog for rangerLogger, you can provide default impl in base so that you don't need to touch the other unrelated classes.
So, for endLog if you see, polymorphism isn't working quite right. If you choose Option 1, I believe that should be fixed and base will behave like base for endLog operation as well.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java (hunk 5 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/state/RangerDumpBegin.java:30]: Why does this annotation required by json deserializer? Is it to deal with a cease to ignore if null?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java (hunk 1 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/log/BootstrapDumpLogger.java]: yes will use option 1
